### DISCUSSION TITLE
Is there a way to force the use of a field alias in serialization even when `model_dump(by_alias=False)`?

### DISCUSSION THREAD

**Discussion by LunkRat** [pydantic/pydantic#8992] [Question]
Hi, I'm using Pydantic v2 and I have a model hierarchy like this:

```python
class OtherThingAttachedToMainThing(BaseModel):
    from_: Optional[str] = Field(default=None, alias="from")
```

```python
class MainThing(BaseModel):
    dump_this_name: OtherThingAttachedToMainThing = Field(default=None, alias="not_dump_this_name")
```

Now, I need to do this:

```python
main_thing.model_dump(by_alias=False)
```

But I want the `from_` field to dump with alias `from` while at the same time I want the `dump_this_name` field (and dozens of others on the MainThing model) to dump with field names not aliases. 

In other words, is there a way to get a single field to ignore `by_alias=False`?

Alternatively (for me) is there a way to dump the parent `MainThing` with `by_alias=False` but have its nested `OtherThingAttachedToMainThing` model dump with `by_alias=True`?

I've attempted `@field_serializer` and `@model_serializer` decorators, and I've explored model config parameters, and `Field()` config parameters, but so far I can't see how to do this.

**Comment by mvanderlee** [comment]
The best way I've found is to use a custom `model_serializer`

```python
    @model_serializer(when_used="always", mode="wrap")
    def clean_model(
        self, serializer: SerializerFunctionWrapHandler, info: SerializationInfo,
    ) -> dict[str, Any]:
        data: dict[str, Any] = serializer(self)
        data['from'] = data.pop('from_')
        return data
```

**Comment by vinicq** [comment]
A `field_serializer` on the parent is the lightest version of this. You don't need to wrap the whole `MainThing` serializer just to flip aliasing on one nested field.

```python
from typing import Optional
from pydantic import BaseModel, Field, field_serializer


class OtherThingAttachedToMainThing(BaseModel):
    from_: Optional[str] = Field(default=None, alias="from")


class MainThing(BaseModel):
    dump_this_name: OtherThingAttachedToMainThing = Field(
        default=None, alias="not_dump_this_name"
    )

    @field_serializer("dump_this_name")
    def _serialize_inner(self, value: OtherThingAttachedToMainThing):
        return value.model_dump(by_alias=True)
```

Then on the parent:

```python
main_thing.model_dump(by_alias=False)
```

The parent uses field names, the inner one uses aliases. `mvanderlee`'s `model_serializer` wrap is the right tool when you need to touch multiple fields or do anything more involved than "serialize this nested model with a different setting." For the single-field case `field_serializer` keeps the intent local.

Heads up that if `OtherThing` itself has nested aliased fields, `by_alias` propagates downward through `model_dump`, so you don't need to repeat the pattern on the inner models. The flip only happens at the explicit boundary.

**Comment by ltianyi992** [comment]
Yes — use `serialization_alias` on the field you want to always serialize under a fixed name, regardless of the parent's `by_alias` setting. This is different from `alias` (which controls both validation and serialization).

```python
from typing import Optional
from pydantic import BaseModel, Field

class OtherThingAttachedToMainThing(BaseModel):
    model_config = {"populate_by_name": True}
    from_: Optional[str] = Field(
        default=None,
        alias="from",
        serialization_alias="from",  # always serializes as "from"
    )

class MainThing(BaseModel):
    dump_this_name: OtherThingAttachedToMainThing = Field(
        default=None,
        alias="not_dump_this_name"
    )

obj = MainThing(**{"not_dump_this_name": {"from": "hello"}})

# by_alias=False on parent: dump_this_name uses field name,
# but from_ still uses serialization_alias="from"
print(obj.model_dump(by_alias=False))
# {'dump_this_name': {'from': 'hello'}}
```

**How this works:**
- `alias` controls what name is accepted on input (validation)
- `serialization_alias` controls the output name and overrides `by_alias` — it is *always* used when set, regardless of whether the caller passes `by_alias=True` or `by_alias=False`

So `serialization_alias="from"` on `from_` guarantees that field always appears as `from` in `model_dump()` output, while other fields on the parent model respect the `by_alias` flag normally.

**Comment by s-chudmunge** [comment]
The cleanest solution for your exact case — a single nested field that must always use its alias regardless of the parent's `by_alias` setting — is `serialization_alias`:

```python
from typing import Optional
from pydantic import BaseModel, Field

class OtherThingAttachedToMainThing(BaseModel):
    model_config = {"populate_by_name": True}
    from_: Optional[str] = Field(
        default=None,
        alias="from",
        serialization_alias="from",  # always emits "from", ignores by_alias
    )

class MainThing(BaseModel):
    dump_this_name: OtherThingAttachedToMainThing = Field(
        default=None, alias="not_dump_this_name"
    )

obj = MainThing(**{"not_dump_this_name": {"from": "hello"}})

# Parent uses field names (by_alias=False), inner field still uses "from"
print(obj.model_dump(by_alias=False))
# {'dump_this_name': {'from': 'hello'}}
```

**Why this works:**

- `alias` controls what name is accepted during *validation* (input).
- `serialization_alias` controls the output name during *serialization* and takes priority over `by_alias` — when set, it is **always** used regardless of whether the caller passes `by_alias=True` or `by_alias=False`.
- `populate_by_name=True` on the inner model lets it accept both `"from"` and `"from_"` as input, which is usually what you want alongside a `serialization_alias`.

---

**When you need more control** (e.g. flip aliasing at the nested model boundary rather than per-field), `field_serializer` on the parent is the right tool:

```python
from pydantic import BaseModel, Field, field_serializer

class OtherThingAttachedToMainThing(BaseModel):
    from_: Optional[str] = Field(default=None, alias="from")

class MainThing(BaseModel):
    dump_this_name: OtherThingAttachedToMainThing = Field(
        default=None, alias="not_dump_this_name"
    )

    @field_serializer("dump_this_name")
    def _serialize_inner(self, value: OtherThingAttachedToMainThing):
        return value.model_dump(by_alias=True)
```

This lets the parent dump by field name while the nested model dumps by alias. Prefer `serialization_alias` for the single-field case; `field_serializer` when you need to override behaviour across an entire nested model.
